lib/key/utils.js

```diff
@@ -99,6 +99,73 @@ export async function cloneKey(inputKey) {
     return key;
 }
 
+/**
+ * Generate a forwarding key for the final recipient, as well as the corresponding proxy factor.
+ * The key in input must be a ECC primary key and must have at least one ECDH subkey using curve25519
+ * @param {Key}               originalKey       ECC primary key of original recipient
+ * @param {Array<Object>}     forwardingUserIds array of user IDs of forwarding recipient
+ * @param {module:type/keyid} subkeyId          (optional) keyid of the ECDH subKey to use for the original recipient
+ * @returns {Promise<Object>}                   The generated key object in the form:
+ *          { proxyFactor, finalRecipientKey: { key:Key, privateKeyArmored:String, publicKeyArmored:String, revocationCertificate:String } }
+ * @async
+ * @static
+ */
+export async function generateForwardingMaterial(originalKey, forwardingUserIds, subKeyId = null) {
+    const curveName = 'curve25519';
+
+    let forwardingKey = await cloneKey(originalKey);
+    // Generate new primary key params
+    await forwardingKey.keyPacket.generate(null, forwardingKey.keyPacket.params[0]);
+    forwardingKey.keyPacket.keyid = null;
+    forwardingKey.keyPacket.fingerprint = null;
+
+    // Setup subKey: find ECDH subkey to override
+    const originalSubKey = await originalKey.getEncryptionKey(subKeyId);
+    if (
+        !originalSubKey ||
+        originalSubKey.getAlgorithmInfo().algorithm !== 'ecdh' ||
+        originalSubKey.keyPacket.params[0].getName() !== curveName
+    ) {
+        throw new Error('Could not find a suitable ECDH encryption key packet');
+    }
+    // Discard all other copied subkeys
+    forwardingKey.subKeys = forwardingKey.subKeys.filter((subKey) =>
+        subKey.getKeyId().equals(originalSubKey.getKeyId())
+    );
```

Comment:
I think we still have to give this some thought. Once we start doing automatic key rotation; maybe we want to create forwarding keys for multiple subkeys? Can we use the same proxy factor for each of them? Etc.